### ISSUE TITLE
Enable Specifying Dependencies of `serve`

### ISSUE THREAD

**Issue by cramertj** [hyperium/hyper#1704]
Currently, [several bounds of the `serve` function](https://github.com/hyperium/hyper/blob/master/src/server/mod.rs#L339) rely on private traits and types which aren't exposed from `hyper`. This makes it impossible to create a function in another library which accepts a compatible `Executor` type and passes it onto the `serve` function. Would it be possible to either expose these traits, or expose the [`Exec enum`](https://github.com/hyperium/hyper/blob/master/src/common/exec.rs#L22) which implements them, so that libraries which wrap hyper can pass on this dependency to their users?

**Comment by seanmonstar** [comment]
Which bounds specifically? The exec traits? Are they not satisfied with `Executor` bounds?

Are you trying to be generic over the executor also? Since there is `E = Exec` on the structs, you should be able to use it by just omitting the second generic, right?

**Comment by cramertj** [comment]
> Are they not satisfied with `Executor` bounds?

`Executor` in 0.1 is generic on the type of future it can execute, so there's no way to request and `Executor` that can execute the required (private) future type (`H2Stream<F, B>`).

> Are you trying to be generic over the executor also? Since there is E = Exec on the structs, you should be able to use it by just omitting the second generic, right?

Yes, the purpose of my question was to allow callers to specify their own executor.

**Comment by seanmonstar** [comment]
Looking at those traits some more, I'm reminded why they aren't public: they expose a bunch of internal pieces, like the watcher stuff. A user shouldn't need to care about any of them, and I'd prefer they didn't have to be part of the trait, but without them, it was impossible to properly propagate the "sendiness" of the user's futures and the executor. If *that* could be done more easily, these weird traits wouldn't be needed at all....

**Comment by cramertj** [comment]
Yeah, ideally you'd have something like `for<T: Future<Output = ()> + Send> Executor<T>`, but alas the trait system isn't there yet.
